File: 0125-Valid-Palindrome/Solution.py
# Completed on: 9/15/2024
# Submission: https://leetcode.com/problems/valid-palindrome/submissions/1391605089/
# Time Complexity: O(n)
# Space Complexity: O(1) 

class Solution:
    def isPalindrome(self, s:str)-> bool:
        l = 0
        r = len(s) - 1
        s  = s.lower()


        while l <= r:
            if not s[l].isalnum():
                l += 1
                continue
            elif not s[r].isalnum():
                r -=1
                continue
            elif s[l] != s[r]:
                return False
            else:
                l+=1
                r-=1
        return True
    

File: 0125-Valid-Palindrome/Solution.cpp
// Completed on: 15/9/2024
// Submission: https://leetcode.com/problems/valid-palindrome/submissions/1391626737/
// Time Complexity: O(n)
// Space Complexity: O(1) 


#include <string>
using namespace std;

class Solution{
    public:

    bool isPalindrome(string s){
        int left = 0;
        int right = s.length() -1;
        string lowered = "";
        for (int i = 0 ; i < s.length() ; i++){
            lowered += tolower(s[i]);
        }
        s = lowered;
        

        while(left <= right){
            if (!isalnum(s[left]))
            {
                left++;
                continue;
            }
            else if (!isalnum(s[right])){
                right--;
                continue;
            }
            else if (s[left] != s[right]){
                return false;
            }
            left ++;
            right --;
        }
        return true;
    }
};